### PR TITLE
Correct CROMFS hardlink handling

### PR DESCRIPTION
**Pull request by davids5**
## Summary

This abstracts away the the hard link and treats them as directories. The internal data representation is not returned if the tyle was formally a DIRLINK.

## Impact

Fixes master that has been broken since  [vfs/stat: Make the flag defintion more confirm POSIX standard.](https://github.com/apache/incubator-nuttx/pull/1522) came in.

## Testing
```
   nsh> mount
     /etc type cromfs
     /fs/microsd type vfat
     /proc type procfs

   test Commands:
     ls -Rl /etc
     ls -l /etc/init.d
     ls -l /etc/init.d/.
     ls -l /etc/init.d/..
     ls -l /etc/init.d/../.

   nsh> ls -Rl /etc
   /etc:
    dr-xr-xr-x       0 .
    dr-xr-xr-x       0 init.d/
    dr-xr-xr-x       0 extras/
    dr-xr-xr-x       0 mixers/
   /etc/init.d:
    dr-xr-xr-x       0 .
    dr-xr-xr-x       0 ..
    -r--r--r--      59 rc.uuv_apps
    -r--r--r--    6495 rcS
    -r--r--r--     794 rc.fw_defaults
    -r--r--r--     265 rc.boat_defaults
    -r--r--r--      85 rc.board_defaults
    -r--r--r--    7603 rc.interface
    -r--r--r--     207 rc.io
    dr-xr-xr-x       0 airframes/
    -r--r--r--     110 rc.fw_apps
    -r--r--r--     780 rc.serial_port
    -r--r--r--     105 rc.autostart.post
    -r--r--r--     538 rc.mc_apps
    -r--r--r--    1446 rc.sensors
    -r--r--r--     786 rc.thermal_cal
    -r--r--r--    6164 rc.autostart
    -r--r--r--     265 rc.rover_defaults
    -r--r--r--    1576 rc.vehicle_setup
    -r--r--r--      61 rc.rover_apps
    -r--r--r--    3319 rc.serial
    -r--r--r--     372 rc.logging
    -r--r--r--     249 rc.vtol_apps
    -r--r--r--      30 rc.board_mavlink
    -r--r--r--     296 rc.mc_defaults
    -r--r--r--     137 rc.board_sensors
    -r--r--r--     497 rc.vtol_defaults
    -r--r--r--     240 rc.uuv_defaults
   /etc/init.d/airframes:
    dr-xr-xr-x       0 .
    dr-xr-xr-x       0 ..
    -r--r--r--     754 4050_generic_250
    -r--r--r--     861 50002_traxxas_stampede_2wd
    -r--r--r--     593 3031_phantom
    -r--r--r--     605 3030_io_camflyer
    -r--r--r--     990 50004_dfrobot_gpx_asurada
    -r--r--r--      67 14002_tri_y_yaw-
    -r--r--r--    2143 13014_vtol_babyshark
    -r--r--r--    1061 13005_vtol_AAERT_quad
    -r--r--r--     117 4100_tiltquadrotor
    -r--r--r--      45 3035_viper
    -r--r--r--      84 1001_rc_quad_x.hil
    -r--r--r--     388 2106_albatross
    -r--r--r--     219 13200_generic_vtol_tailsitter
    -r--r--r--      55 3000_generic_wing
    -r--r--r--     564 3033_wingwing
    -r--r--r--      67 7001_hexa_+
    -r--r--r--     750 4080_zmr250
    -r--r--r--      67 8001_octo_x
    -r--r--r--     660 13001_caipirinha_vtol
    -r--r--r--     871 13006_vtol_standard_delta
    -r--r--r--     276 13050_generic_vtol_octo
    -r--r--r--     857 4052_holybro_qav250
    -r--r--r--    1027 3100_tbs_caipirinha
    -r--r--r--     154 50001_axialracing_ax10
    -r--r--r--     814 13002_firefly6
    -r--r--r--     445 4060_dji_matrice_100
    -r--r--r--      67 9001_octo_+
    -r--r--r--     272 4014_s500
    -r--r--r--     585 4040_reaper
    -r--r--r--     441 10017_steadidrone_qu4d
    -r--r--r--     882 4051_s250aq
    -r--r--r--     185 3034_fx79
    -r--r--r--      63 4001_quad_x
    -r--r--r--     442 4010_dji_f330
    -r--r--r--     266 13003_quad_tailsitter
    -r--r--r--     662 1000_rc_fw_easystar.hil
    -r--r--r--     490 3032_skywalker_x5
    -r--r--r--     442 4011_dji_f450
    -r--r--r--     424 10015_tbs_discovery
    -r--r--r--    1090 50003_aion_robotics_r1_rover
    -r--r--r--      67 14001_tri_y_yaw+
    -r--r--r--    1232 1002_standard_vtol.hil
    -r--r--r--     481 10016_3dr_iris
    -r--r--r--     555 4003_qavr5
    -r--r--r--     262 13000_generic_vtol_standard
    -r--r--r--      69 12001_octo_cox
    -r--r--r--     741 15001_coax_heli
    -r--r--r--      63 5001_quad_+
    -r--r--r--     465 4500_clover4
    -r--r--r--     394 2200_mini_talon
    -r--r--r--     510 17002_TF-AutoG2
    -r--r--r--    2513 4016_holybro_px4vision
    -r--r--r--     928 4053_holybro_kopis2
    -r--r--r--     866 50000_generic_ground_vehicle
    -r--r--r--    3240 13013_deltaquad
    -r--r--r--     180 60001_uuv_hippocampus
    -r--r--r--     339 4031_3dr_quad
    -r--r--r--     597 12002_steadidrone_mavrik
    -r--r--r--     501 13010_claire
    -r--r--r--    1476 13012_convergence
    -r--r--r--     740 16001_helicopter
    -r--r--r--     387 2105_maja
    -r--r--r--     329 4015_holybro_s500
    -r--r--r--      84 1100_rc_quad_x_sih.hil
    -r--r--r--     612 4009_qav250
    -r--r--r--     180 60000_uuv_generic
    -r--r--r--     621 3036_pigeon
    -r--r--r--     519 10018_tbs_endurance
    -r--r--r--     791 13007_vtol_AAVVT_quad
    -r--r--r--     241 13004_quad+_tailsitter
    -r--r--r--      67 6001_hexa_x
    -r--r--r--     617 3037_parrot_disco_mod
    -r--r--r--      69 11001_hexa_cox
    -r--r--r--    1723 13009_vtol_spt_ranger
    -r--r--r--    1176 13008_QuadRanger
    -r--r--r--     124 2100_standard_plane
    -r--r--r--     440 24001_dodeca_cox
    -r--r--r--     110 rc.fw_apps
    -r--r--r--     780 rc.serial_port
    -r--r--r--     105 rc.autostart.post
    -r--r--r--     538 rc.mc_apps
    -r--r--r--    1446 rc.sensors
    -r--r--r--     786 rc.thermal_cal
    -r--r--r--    6164 rc.autostart
    -r--r--r--     265 rc.rover_defaults
    -r--r--r--    1576 rc.vehicle_setup
    -r--r--r--      61 rc.rover_apps
    -r--r--r--    3319 rc.serial
    -r--r--r--     372 rc.logging
    -r--r--r--     249 rc.vtol_apps
    -r--r--r--      30 rc.board_mavlink
    -r--r--r--     296 rc.mc_defaults
    -r--r--r--     137 rc.board_sensors
    -r--r--r--     497 rc.vtol_defaults
    -r--r--r--     240 rc.uuv_defaults
   /etc/extras:
    dr-xr-xr-x       0 .
    dr-xr-xr-x       0 ..
    -r-xr-xr-x   10448 px4fmuv3_bl.bin
    -r-xr-xr-x   56692 px4_io-v2_default.bin
    dr-xr-xr-x       0 mixers/
   /etc/extras/mixers:
    dr-xr-xr-x       0 .
    dr-xr-xr-x       0 ..
    -r--r--r--     224 tilt_quad.aux.mix
    -r--r--r--     230 coax.main.mix
    -r--r--r--      26 firefly6.main.mix
    -r--r--r--     104 quad_h.main.mix
    -r--r--r--      26 hexa_+.main.mix
    -r--r--r--     197 mount_legs.aux.mix
    -r--r--r--     104 quad_+.main.mix
    -r--r--r--      25 dodeca_bottom_cox.aux.mix
    -r--r--r--      26 octo_+.main.mix
    -r--r--r--      26 octo_x.main.mix
    -r--r--r--     148 generic_diff_rover.main.mix
    -r--r--r--      27 octo_cox_w.main.mix
    -r--r--r--     236 vtol_tailsitter_duo.main.mix
    -r--r--r--     156 mount.aux.mix
    -r--r--r--     245 deltaquad.main.mix
    -r--r--r--      84 rover_generic.main.mix
    -r--r--r--     336 vtol_AAVVT.aux.mix
    -r--r--r--      27 quad_dc.main.mix
    -r--r--r--      28 quad_x_cw.main.mix
    -r--r--r--      26 hexa_cox.main.mix
    -r--r--r--     183 delta.main.mix
    -r--r--r--     104 quad_w.main.mix
    -r--r--r--     183 caipi.main.mix
    -r--r--r--     300 Viper.main.mix
    -r--r--r--     156 pass.aux.mix
    -r--r--r--     179 claire.aux.mix
    -r--r--r--     183 FX79.main.mix
    -r--r--r--     312 IO_pass.main.mix
    -r--r--r--     564 uuv_x.main.mix
    -r--r--r--      84 stampede.main.mix
    -r--r--r--     170 quad_x_vtol.main.mix
    -r--r--r--     259 vtol_delta.aux.mix
    -r--r--r--     104 quad_s250aq.main.mix
    -r--r--r--      26 octo_cox.main.mix
    -r--r--r--     191 blade130.main.mix
    -r--r--r--     296 tilt_quad.main.mix
    -r--r--r--     251 babyshark.main.mix
    -r--r--r--     187 phantom.main.mix
    -r--r--r--     110 quad_x.main.mix
    -r--r--r--     187 wingwing.main.mix
    -r--r--r--     103 zmr250.main.mix
    -r--r--r--      26 claire.main.mix
    -r--r--r--     313 vtol_convergence.main.mix
    -r--r--r--     438 TF-AutoG2.main.mix
    -r--r--r--      26 hexa_x.main.mix
    -r--r--r--      25 dodeca_top_cox.main.mix
    -r--r--r--      65 tri_y_yaw+.main.mix
    -r--r--r--      29 ocpoc_quad_x.main.mix
    -r--r--r--     219 firefly6.aux.mix
    -r--r--r--     451 AAVVTWFF_vtail.main.mix
    -r--r--r--     172 rover_diff_and_servo.main.mix
    -r--r--r--     256 quad_+_vtol.main.mix
    -r--r--r--     451 AAVVTWFF.main.mix
    -r--r--r--     266 AETRFG.main.mix
    -r--r--r--     185 standard_vtol_hitl.main.mix
    -r--r--r--     187 fw_generic_wing.main.mix
    -r--r--r--      67 tri_y_yaw-.main.mix
    -r--r--r--     368 CCPM.main.mix
    -r--r--r--     265 AERT.main.mix
    -r--r--r--     194 vtol_AAERT.aux.mix
    -r--r--r--     424 AAERTWF.main.mix
   /etc/mixers:
    dr-xr-xr-x       0 .
    dr-xr-xr-x       0 ..
    -r--r--r--     224 tilt_quad.aux.mix
    -r--r--r--     230 coax.main.mix
    -r--r--r--      26 firefly6.main.mix
    -r--r--r--     104 quad_h.main.mix
    -r--r--r--      26 hexa_+.main.mix
    -r--r--r--     197 mount_legs.aux.mix
    -r--r--r--     104 quad_+.main.mix
    -r--r--r--      25 dodeca_bottom_cox.aux.mix
    -r--r--r--      26 octo_+.main.mix
    -r--r--r--      26 octo_x.main.mix
    -r--r--r--     148 generic_diff_rover.main.mix
    -r--r--r--      27 octo_cox_w.main.mix
    -r--r--r--     236 vtol_tailsitter_duo.main.mix
    -r--r--r--     156 mount.aux.mix
    -r--r--r--     245 deltaquad.main.mix
    -r--r--r--      84 rover_generic.main.mix
    -r--r--r--     336 vtol_AAVVT.aux.mix
    -r--r--r--      27 quad_dc.main.mix
    -r--r--r--      28 quad_x_cw.main.mix
    -r--r--r--      26 hexa_cox.main.mix
    -r--r--r--     183 delta.main.mix
    -r--r--r--     104 quad_w.main.mix
    -r--r--r--     183 caipi.main.mix
    -r--r--r--     300 Viper.main.mix
    -r--r--r--     156 pass.aux.mix
    -r--r--r--     179 claire.aux.mix
    -r--r--r--     183 FX79.main.mix
    -r--r--r--     312 IO_pass.main.mix
    -r--r--r--     564 uuv_x.main.mix
    -r--r--r--      84 stampede.main.mix
    -r--r--r--     170 quad_x_vtol.main.mix
    -r--r--r--     259 vtol_delta.aux.mix
    -r--r--r--     104 quad_s250aq.main.mix
    -r--r--r--      26 octo_cox.main.mix
    -r--r--r--     191 blade130.main.mix
    -r--r--r--     296 tilt_quad.main.mix
    -r--r--r--     251 babyshark.main.mix
    -r--r--r--     187 phantom.main.mix
    -r--r--r--     110 quad_x.main.mix
    -r--r--r--     187 wingwing.main.mix
    -r--r--r--     103 zmr250.main.mix
    -r--r--r--      26 claire.main.mix
    -r--r--r--     313 vtol_convergence.main.mix
    -r--r--r--     438 TF-AutoG2.main.mix
    -r--r--r--      26 hexa_x.main.mix
    -r--r--r--      25 dodeca_top_cox.main.mix
    -r--r--r--      65 tri_y_yaw+.main.mix
    -r--r--r--      29 ocpoc_quad_x.main.mix
    -r--r--r--     219 firefly6.aux.mix
    -r--r--r--     451 AAVVTWFF_vtail.main.mix
    -r--r--r--     172 rover_diff_and_servo.main.mix
    -r--r--r--     256 quad_+_vtol.main.mix
    -r--r--r--     451 AAVVTWFF.main.mix
    -r--r--r--     266 AETRFG.main.mix
    -r--r--r--     185 standard_vtol_hitl.main.mix
    -r--r--r--     187 fw_generic_wing.main.mix
    -r--r--r--      67 tri_y_yaw-.main.mix
    -r--r--r--     368 CCPM.main.mix
    -r--r--r--     265 AERT.main.mix
    -r--r--r--     194 vtol_AAERT.aux.mix
    -r--r--r--     424 AAERTWF.main.mix
   nsh> ls -l /etc/init.d
   /etc/init.d:
    dr-xr-xr-x       0 .
    dr-xr-xr-x       0 ..
    -r--r--r--      59 rc.uuv_apps
    -r--r--r--    6495 rcS
    -r--r--r--     794 rc.fw_defaults
    -r--r--r--     265 rc.boat_defaults
    -r--r--r--      85 rc.board_defaults
    -r--r--r--    7603 rc.interface
    -r--r--r--     207 rc.io
    dr-xr-xr-x       0 airframes/
    -r--r--r--     110 rc.fw_apps
    -r--r--r--     780 rc.serial_port
    -r--r--r--     105 rc.autostart.post
    -r--r--r--     538 rc.mc_apps
    -r--r--r--    1446 rc.sensors
    -r--r--r--     786 rc.thermal_cal
    -r--r--r--    6164 rc.autostart
    -r--r--r--     265 rc.rover_defaults
    -r--r--r--    1576 rc.vehicle_setup
    -r--r--r--      61 rc.rover_apps
    -r--r--r--    3319 rc.serial
    -r--r--r--     372 rc.logging
    -r--r--r--     249 rc.vtol_apps
    -r--r--r--      30 rc.board_mavlink
    -r--r--r--     296 rc.mc_defaults
    -r--r--r--     137 rc.board_sensors
    -r--r--r--     497 rc.vtol_defaults
    -r--r--r--     240 rc.uuv_defaults
   nsh> ls -l /etc/init.d/.
   /etc/init.d/.:
    dr-xr-xr-x       0 .
    dr-xr-xr-x       0 ..
    -r--r--r--      59 rc.uuv_apps
    -r--r--r--    6495 rcS
    -r--r--r--     794 rc.fw_defaults
    -r--r--r--     265 rc.boat_defaults
    -r--r--r--      85 rc.board_defaults
    -r--r--r--    7603 rc.interface
    -r--r--r--     207 rc.io
    dr-xr-xr-x       0 airframes/
    -r--r--r--     110 rc.fw_apps
    -r--r--r--     780 rc.serial_port
    -r--r--r--     105 rc.autostart.post
    -r--r--r--     538 rc.mc_apps
    -r--r--r--    1446 rc.sensors
    -r--r--r--     786 rc.thermal_cal
    -r--r--r--    6164 rc.autostart
    -r--r--r--     265 rc.rover_defaults
    -r--r--r--    1576 rc.vehicle_setup
    -r--r--r--      61 rc.rover_apps
    -r--r--r--    3319 rc.serial
    -r--r--r--     372 rc.logging
    -r--r--r--     249 rc.vtol_apps
    -r--r--r--      30 rc.board_mavlink
    -r--r--r--     296 rc.mc_defaults
    -r--r--r--     137 rc.board_sensors
    -r--r--r--     497 rc.vtol_defaults
    -r--r--r--     240 rc.uuv_defaults
   nsh> ls -l /etc/init.d/..
   /etc/init.d/..:
    dr-xr-xr-x       0 .
    dr-xr-xr-x       0 init.d/
    dr-xr-xr-x       0 extras/
    dr-xr-xr-x       0 mixers/
   nsh> ls -l /etc/init.d/../.
   /etc/init.d/../.:
    dr-xr-xr-x       0 .
    dr-xr-xr-x       0 init.d/
    dr-xr-xr-x       0 extras/
    dr-xr-xr-x       0 mixers/
```

### CODE REVIEW

**Comment by patacongo** [comment]
I am going to close this now.  It is not correct and should not be merged.  I intend to implement the correct following of hardlinks in the next week or so.  Changing the flags is not the correct solution.

**Comment by davids5** [comment]
@patacongo - Perfect - The fix will wait for you.

File: tools/gencromfs.c [hunk 1]
```diff
@@ -111,7 +111,7 @@
 #define NUTTX_IFLNK        (10 << 12)  /* Must match NuttX's S_IFLNK */
 
 #define DIR_MODEFLAGS      (NUTTX_IFDIR | NUTTX_IRXUSR | NUTTX_IRXGRP | NUTTX_IRXOTH)
-#define DIRLINK_MODEFLAGS  (NUTTX_IFLNK | DIR_MODEFLAGS)
+#define DIRLINK_MODEFLAGS  (NUTTX_IFLNK | NUTTX_IRXUSR | NUTTX_IRXGRP | NUTTX_IRXOTH)
```
Comment (patacongo) [inline tools/gencromfs.c:114]: This part of the PR looks correct.  The rest is wrong.  Can we commit this separately?

Comment (davids5) [reply tools/gencromfs.c:114]: Done see https://github.com/apache/incubator-nuttx/pull/1551

Comment (patacongo) [reply tools/gencromfs.c:114]: Okay... I created #1552.  I will close #1552

**Pull request closed without merging**